app/models/rule.rb

```diff
@@ -22,4 +22,16 @@ class Rule < ApplicationRecord
   validates :text, presence: true, length: { maximum: TEXT_SIZE_LIMIT }
 
   scope :ordered, -> { kept.order(priority: :asc, id: :asc) }
+
+  def move(offset)
+    rules = Rule.ordered.to_a
+    position = rules.index(self)
+
+    rules.delete_at(position)
+    rules.insert(position + offset, self)
+
+    rules.each.with_index do |rule, index|
+      rule.update(priority: index)
+    end
```

Comment:
How about wrapping this in a transaction (and using `#update!` for good measure)? This way either all positions get updated or none and no inconsistent state can be created.